### PR TITLE
7593 new AtLeastValidator

### PR DESCRIPTION
**Pull request by sdlins**

### CODE REVIEW

**Comment by slinstj** [comment]
Have these tests fails to do with some unavailability of the Travis CI?

**Comment by slinstj** [comment]
#7593

**Comment by unclead** [comment]
I found that you have invalid name of test class "BooleanValidatorTest" instead  of AtLeastOneValidatorTest

**Comment by slinstj** [comment]
Thank you. I will fix! I used that one as base.

**Comment by slinstj** [comment]
`AtLeastOne` >> `AtLeast`
`validateWith` >> `with`
`errorAttributes` >> `errorIn`
Added `min` param

**Comment by Faryshta** [comment]
How about adding a new parameter `RequiredValidator::targetAttribute`?

**Comment by SilverFire** [comment]
@Faryshta instead of this PR?

**Comment by slinstj** [comment]
I think a `targetAttribute` in RequiredValidator would be messy. The attributes are *not* required, they are *alternative* between them.

**Comment by Faryshta** [comment]
@SilverFire could be an option, i see this PR has been inactive for sometime. I am cool with a new vaildator or adding functionality to an existing one.

@slinstj well to me it sounds like one of them is required. Again, its just an alternative since this PR seems inactive.

**Comment by SilverFire** [comment]
@Faryshta I think it's better to create a new validator instead of overcomplicating existing one

**Comment by Faryshta** [comment]
ok cool, what does this need to be accepted?

File: framework/validators/AtLeastValidator.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * Important: Only one attribute should be specified in first param of the rule.
+ * The aditional ones must go into [[with]] param.
+ *
+ * The attributes that will receive the error message can be specified into [[errorIn]]
+ * param. If none is specified, all envolved attributes will receive the error.
+ *
+ * @author Sidney Lins <slinstj@gmail.com>
+ * @since 2.0.4
+ */
+class AtLeastValidator extends Validator
```
Comment (cebe) [inline framework/validators/AtLeastValidator.php]: I'm not really happy with the name, "at least" is missing something. at least... what? Alternatives:

- RequireFrom
- RequireOne
- RequireAny
- Alternative
- AtLeastOne
- ...

File: framework/validators/AtLeastValidator.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+{
+    /**
+     * @var array|string the required list of aditional attributes in which to check
+     * at least one filled.
+     */
+    public $with;
```
Comment (cebe) [inline framework/validators/AtLeastValidator.php]: also naming: `alternativeAttributes`, `targetAttributes`, ...

File: framework/validators/AtLeastValidator.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * @var integer the minimun required quantity of attributes that must to be filled.
+     * Defaults to 1.
+     */
+    public $min = 1;
+    /**
+     * @var array the list of attributes that should receive the error message.
+     * Defaults to all attributes being validated (i.e, attribute + [[with]] attributes).
+     */
+    public $errorIn;
```
Comment (cebe) [inline framework/validators/AtLeastValidator.php]: naming: `errorAttributes`

File: framework/validators/AtLeastValidator.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        if (empty($this->with)) {
+            throw new InvalidConfigException('Param "with" can not be null');
+        }
+        if ($this->message === null) {
+            $this->message = Yii::t('yii', 'Please, fill {attributesList}.');
```
Comment (cebe) [inline framework/validators/AtLeastValidator.php]: `Please fill at least {min} of {attributesList}.`

File: framework/validators/AtLeastValidator.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    public function validateAttribute($model, $attribute)
+    {
+        $value = $model->$attribute;
+
+        $attributes = array_merge((array) $attribute, (array) $this->with);
+        $filled = 0;
+        foreach ($attributes as $attribute) {
+            $value = $model->$attribute;
+            $attributesListLabels[] = '"' . $model->generateAttributeLabel($attribute) . '"';
+            $filled += !empty($value) ? 1 : 0;
```
Comment (cebe) [inline framework/validators/AtLeastValidator.php]: better use the `isEmpty()` method of Validator.

File: framework/validators/AtLeastValidator.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        if (!$filled) {
```
Comment (cebe) [inline framework/validators/AtLeastValidator.php]: should compare to `$this->min` here

File: framework/validators/AtLeastValidator.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+ * ~~~
+ *
+ * Important: Only one attribute should be specified in first param of the rule.
+ * The aditional ones must go into [[with]] param.
+ *
+ * The attributes that will receive the error message can be specified into [[errorIn]]
+ * param. If none is specified, all envolved attributes will receive the error.
+ *
+ * @author Sidney Lins <slinstj@gmail.com>
+ * @since 2.0.4
```
Comment (cebe) [inline framework/validators/AtLeastValidator.php]: update to 2.0.13

**Review by cebe — CHANGES_REQUESTED** [review]

**Comment by MaXL-ru** [comment]
I vote for `RequireAny` validator name.

@slinstj can you make changes ? I'll be very glad to see this feature in nearest Yii2 version.

**Comment by slinstj** [comment]
@MaXL-ru Please, go ahead and make the changes (if it is ok to you). I can't do it for now (no time). Sorry.

**Comment by slinstj** [comment]
@developeruz sorry, not sure what you need. I did not lock anything and I can't see any place to 'allow' your changes. Please, explain it better sothat I can help.

**Comment by developeruz** [comment]
@slinstj on the left sidebar of this page, you should see "Allow edits from maintainers" checkbox (please look at the attached image in the previous comment). You and only you, as an owner of PR,  can mark this checkbox. Without that permission, no one except you can modify files in your branch.

**Comment by samdark** [comment]
Replaced by https://github.com/yiisoft/yii2/pull/15157

**Pull request closed without merging**